cfbs/commands.py

```diff
@@ -1000,3 +1001,74 @@ def input_command(args, input_from="cfbs input"):
         files_to_commit.append(input_path)
     config.save()
     return Result(0, do_commit, None, files_to_commit)
+
+
+def set_input_command(name, src_path):
+    config = CFBSConfig.get_instance()
+    module = config.get_module_from_build(name)
+    if module is None:
+        log.error("Module '%s' not found" % name)
+        return 1
+
+    spec = module.get("input")
+    if spec is None:
+        log.error("Module '%s' does not accept input" % name)
+        return 1
+    log.debug("Input spec for module '%s': %s" % (name, pretty(spec)))
+
+    if src_path is None:
+        try:
+            data = json.load(stdin)
+        except json.decoder.JSONDecodeError as e:
+            log.error("Error reading json from stdin: %s" % e)
+            return 1
+    else:
+        data = read_json(src_path)
+        if data is None:
+            log.error("Failed to read json from file '%s'" % src_path)
+            return 1
+
+    assert data is not None
+    log.debug("Input data for module '%s': %s" % (name, pretty(data)))
+
+    def _compare_dict(a, b, ignore={}):
+        assert isinstance(a, dict) and isinstance(b, dict)
+        if set(a.keys()) != set(b.keys()) - ignore:
+            return False
+        """Avoid code duplication by converting the values of the two dicts
+        into two lists in the same order and compare the lists instead"""
+        keys = a.keys()
+        return _compare_list([a[key] for key in keys], [b[key] for key in keys])
+
+    def _compare_list(a, b):
```

Comment:
So this requires that things are in the very same order. I'm not sure how much strict JSON is regarding ordering of object's children, but it feels wrong to me to require the inputs to be in the very same order.

Reply:
Objects does not have to be in order. I added a test to prove that. But lists do, and they should

Reply:
Aren't the inputs in a list?

Reply:
Yep! Imagine the following variables being swapped
```
$ cfbs input create-single-file-with-content
--snip--
What file do you want to create? /tmp/testfile
What content should this file have? Hello CFEngine!
```
i.e.
```
$ cfbs input create-single-file-with-content
--snip--
What content should this file have? Hello CFEngine!
What file do you want to create? /tmp/testfile
```
it would not make much sense for the user.

Reply:
It wouldn't. But when the data comes from a form in MP or generally from some JSON, ordering should not matter.

Reply:
The order of the list defines the order of the prompts and the order of the UI elements in MP. It matters. MP should be working on the list it got from get-input, and should not reorder the elements before running set-input.

Reply:
We could enable module authors to more easily change input (reorder it) and have it work automatically with `cfbs update`, but that's not this command.

Reply:
> should not reorder the elements before running set-input
I don't see a good reason for this to be disallowed/unsupported.

Reply:
The order here communicates how it should be shown to the user and when it comes back, how it was shown to the user.

For example, things coming back in another order could indicate that MP is not showing things correctly, or not handling the input received from the user correctly.

It is important to record what the user saw and what they responded, including the order of things. @larsewi showed one example. Similar situations can apply to the GUI. If labels are bad or confusing, or users don't read them, or module authors made mistakes when filling out the labels or variable names, having an accurate record of what the user saw and entered is valuable.

I would take it one step further and say we should also record whether the user saw the prompt (used the CLI) or whether they saw the label (used MP). However this is a bit more difficult, and we are not doing that currently.

Reply:
I guess I'm thinking about this differently -- the input, to me, is just data. So I would expect the `get-input` command to only take answers, basically as `input_id: value` pairs